### DISCUSSION TITLE
Horizontal/Vertical lines interpolation

### DISCUSSION THREAD

**Discussion by DaMilyutin** [slint-ui/slint#819] [Q&A] [answered]
Hi, I was rendering bunch of horizontal/vertical lines with plotters in sixtyfps via ImageBuffer. 
And I was surprised by effect of variaton of brightness.
![image](https://user-images.githubusercontent.com/44639272/148661243-f4b7b687-a76d-47b3-a12b-c9d4f9b55d64.png)
![image](https://user-images.githubusercontent.com/44639272/148661494-416879ee-7eec-4661-b533-f522a291bfc6.png)

To find where it started, I've recreated example output to file (.png) 
and got this (as seen thru Win 10 Photos)
![image](https://user-images.githubusercontent.com/44639272/148661284-74b9dd09-403e-43a6-9263-01f81f9447cd.png)
![image](https://user-images.githubusercontent.com/44639272/148661537-dc8b17f9-1bf6-4862-b3ad-d5fd040c30a1.png)

NOTE: file size may fiffer from buffer size in first examle.

So I decided to load this file in sixty via @image-url and got this
![image](https://user-images.githubusercontent.com/44639272/148661344-2dcac93c-24c0-4132-984c-12dffdc32e99.png)
![image](https://user-images.githubusercontent.com/44639272/148661564-7d29f57b-6fdf-41ac-8d28-73e4d302c937.png)

So is seems like it is not plotters pecularity but the way image is interpolated.
This douesnot bother me a lot but is surprising. So FYI.

X( I think you can recreate this by using my example imagees from Win 10 Photos and using `@image-url`. )X - upd: no you can't

And I add some code. I removed axes but still get
![image](https://user-images.githubusercontent.com/44639272/148661823-ad8b3b5e-bd5d-4757-997f-4a4da768c37a.png)

```
fn render_plot(width: f32, height: f32) -> sixtyfps::Image {
    let width = width as usize;
    let height = height as usize;
    let mut pixel_buffer = SharedPixelBuffer::new(width, height);
    let size = (pixel_buffer.width() as u32, pixel_buffer.height() as u32);

    let backend = BitMapBackend::with_buffer(pixel_buffer.make_mut_bytes(), size);

    // Plotters requires TrueType fonts from the file system to draw axis text - we skip that for
    // WASM for now.
    #[cfg(target_arch = "wasm32")]
    let backend = wasm_backend::BackendWithoutText { backend };

    let root = backend.into_drawing_area();

    root.fill(&BLACK).unwrap();

    let mut chart = ChartBuilder::on(&root)
        .x_label_area_size(width as f64)
        .y_label_area_size(height as f64)
        .top_x_label_area_size(width as f64)
        .right_y_label_area_size(height as f64)
        .build_cartesian_2d(0f32..width as f32, 0f32..height as f32)
        .unwrap();

    chart
        .configure_mesh()
        .disable_x_mesh()
        .disable_y_mesh()
        .draw()
        .unwrap();

    // horizontal
    // for y in (0..(height / 2)).map(|i| 2.0 * i as f32) {
    //     chart
    //         .draw_series(LineSeries::new(
    //             (0..=1).map(|i| (10. + (i as f32) * (width as f32 - 20.), y)),
    //             ShapeStyle {
    //                 color: WHITE.to_rgba(),
    //                 filled: false,
    //                 stroke_width: 1,
    //             },
    //         ))
    //         .unwrap();
    // }

    // vertical
    for x in (0..(width / 4)).map(|i| 4.0 * i as f32) {
        chart
            .draw_series(LineSeries::new(
                (0..=1).map(|i| (x, 10. + (i as f32) * (height as f32 - 20.))),
                ShapeStyle {
                    color: WHITE.to_rgba(),
                    filled: false,
                    stroke_width: 1,
                },
            ))
            .unwrap();
    }

    root.present().unwrap();

    drop(chart);
    drop(root);
    sixtyfps::Image::from_rgb8(pixel_buffer)
}

```

In .60
```
        VerticalBox {
            width: root.width - MenuView.width;
            height: root.height;
            Image {
                //source: @image-url("D:/tmp/test-V.png");
                source: root.render_plot(parent.width, parent.height);
            }
        }
```

**Comment by DaMilyutin** [comment]
And It definitely has same rendering issue in browser.
I can not prove my point with screenshots...

**Reply by DaMilyutin** [reply]
uploaded images from PC. not screenshots.

![test-H](https://user-images.githubusercontent.com/44639272/148661937-7cdd8439-25a8-4951-bd17-e6a2503a2cfa.png)
![test-V](https://user-images.githubusercontent.com/44639272/148661938-449a2267-3a6c-4beb-a1a1-8be49546e4ea.png)

**Comment by ogoffart** [comment, marked as answer]
Looks like a [Moiré pattern](https://en.wikipedia.org/wiki/Moir%C3%A9_pattern) that could be caused by some scaling somewhere in the pipeline. If the image is not rendered at the exact same size of the source, it is going to be scaled. Another problem would occur if we don't draw the image aligned on the pixel grid, which might also be the case.
I'm also seeing some optical illusions looking at the original images, but i guess you don't mean that problem.

**Reply by DaMilyutin** [reply]
Thanks for the answer. 
I pass image sizes 
```
        VerticalBox {
            width: root.width - MenuView.width;
            height: root.height;
            Image {
                //source: @image-url("D:/tmp/test-V.png");
                source: root.render_plot(parent.width, parent.height);
            }
        }
```
and create buffer of thoose sizes except I convert them from float to usize.

How can I neglect this effect on my side?
Ex. Round/floor sizes in .60?

This is not a big concern for me by now. Just to know workaround for future.

**Reply by tronical** [reply]
If it's scaling related, can you try setting the `image-rendering` property on the `Image`? I'm not sure that fixes it, but I wonder if it changes the appearance in your case: `image-rendering: pixelated;`

**Reply by DaMilyutin** [reply]
Thanks. That did work.
![image](https://user-images.githubusercontent.com/44639272/148767983-bf987581-0f2c-4eef-84ce-a1793b849091.png)

**Reply by DaMilyutin** [reply]
@tronical,
For vertical lines I have strange pattern (notice broken lines)
![image](https://user-images.githubusercontent.com/44639272/148772801-1d473bf2-f4a6-45a2-9ab9-4e678efb33e2.png)

when printing to bmp I have this
![image](https://user-images.githubusercontent.com/44639272/148772888-ecafb0bb-a7c2-4906-a845-6394f8772138.png)

**Reply by DaMilyutin** [reply]
@tronical, can you clarify, please. I'm curious and have a little concerns, too.
Image sizes and origin are float. Thus, Image is always forced to be interpolated?
Can image be sowehow matched to pixel grid without interpolation: pixel-to-pixel map?

**Reply by tronical** [reply]
If the image isn’t scaled then the other property that can produce such an artifact would be incorrect positioning. It’s possible that we aren’t placing the pixels correctly into the center (0,5/0,5). That would be a bug IMO. 

I suppose that your issue doesn’t happen when you use the Qt backend?

**Reply by DaMilyutin** [reply]
@tronical, If I got you right, I have no idea how to switch to Qt backend with sixtyfps. 

...else (if I got you *not* right):
I haven't tried rendering this image in Qt. (I haven't opened Qt-related projects for a year at least.)

**Reply by ogoffart** [reply]
To switch backend at runtime, you can use `SIXTYFPS_BACKEND=Qt` or `SIXTYFPS_BACKEND=GL` env variables
Note that the Qt backend is normaly the default if Qt is installed.

Anyway, we currently do not make any effort to draw the images aligned on the pixel grid, with any backend. As far as i'm aware. So bluriness could porentially be reproduced with any backend.  We do need to fix that.
But that doesn't explain the broken lines though.

**Reply by DaMilyutin** [reply]
@ogoffart, Installed Qt just now. Will try. 
Where do I write env variables? I never did that.

**Reply by ogoffart** [reply]
In your shell.
assuming you are using windows, it is something like
`set SIXTYFPS_BACKEND=Qt`
(i think)
But since Qt is the default backend anyway (if it is installed) you don't need to do that

**Reply by DaMilyutin** [reply]
@ogoffart , tried it. See no difference. Pattern is broken.
I'll retry it in my newer C++ playground when install plotting library there.
However, I think vertical lines are good test case. 
Would you create ticket (to be postponed for later but must be tracked) or wait for my newer example?
I plan to do both horizontal and vertical lines subplots and maybe grid.
